tests/test_scenarios.py

```diff
@@ -37,53 +48,119 @@ def test_scenario(filename, monkeypatch):
-
-    if os.getenv("UPDATE_SCENARIOS"):
-        save_scenario(path, cleaned_output)
-    else:
-        assert output == cleaned_output
+        for command in commands:
+            if command.command[:6] == "$ ick ":
+                # TODO: handle global options like -vv
+                args = shlex.split(command.command[6:])
+                result = runner.invoke(main, args, catch_exceptions=False)
+                output = result.output
+            elif command.command == "":
+                output = ""
+            else:
+                assert command.command[:2] == "$ "
+                proc = subprocess.run(
+                    command.command[2:],
+                    encoding="utf-8",
+                    shell=True,
+                    check=False,
+                    stdout=subprocess.PIPE,
+                    stderr=subprocess.STDOUT,
+                )
+                output = proc.stdout
+
+            cleaned_output = clean_output(output)
+            if update:
+                command.output = cleaned_output  # pragma: nocover
+            else:
+                assert command.output == cleaned_output
+
+    if update:  # pragma: nocover
+        save_scenario(path, commands)
+
+
+@dataclasses.dataclass
+class ScenarioCommand:
+    comments: str = ""
+    command: str = ""
+    output: str = ""
 
 
-def load_scenario(path: Path) -> Tuple[Tuple[str, ...], str]:
-    command: Optional[Tuple[str, ...]] = None
-    output: str = ""
-    state = 0
-    with open(path) as f:
-        for line in f:
-            if state == 0 and line.startswith("#"):
-                pass
-            elif state == 0 and line.startswith("$"):
-                parts = shlex.split(line[1:])
-                assert parts[0] == "ick"
-                command = tuple(parts[1:])
-                state = 1
-            elif state == 1:
-                output += line
-
-    assert state == 1
-    assert command is not None
-    return (command, output)
-
-
-def save_scenario(path: Path, new_output: str) -> None:
-    state = 0
-    buf = ""
+def load_scenario(path: Path) -> Iterable[ScenarioCommand]:
     with open(path) as f:
-        for line in f:
-            if state == 0 and line.startswith("#"):
-                buf += line
-            elif state == 0 and line.startswith("$"):
-                buf += line
-                state = 1
-
-    assert state == 1
-    if not buf.endswith("\n"):
-        buf += "\n"
-    path.write_text(buf + new_output)
+        return parse_scenario(f)
+
+
+def parse_scenario(lines: Iterable[str]) -> Iterable[ScenarioCommand]:
+    commands: list[ScenarioCommand] = []
+    command: ScenarioCommand = ScenarioCommand()
+    found_command = False
+
+    def new_command():
+        nonlocal command, found_command
+        if command is not None:
+            commands.append(command)
+        command = ScenarioCommand()
+        found_command = False
+
+    for line in lines:
```

Comment:
If you made a big scary regex like

```
(?P<comments>(?:#.*\n)*))
(?P<command>(?:\$.*\n))
(?P<output>(?:[^#$].*\n|\n)*)
```

I think you wouldn't need `new_command` anymore.